### PR TITLE
Extract `rustc_ast_lowering` crate from `rustc`

### PR DESCRIPTION
**Pull request by Centril**
Working towards https://github.com/rust-lang/rust/issues/65031.

This PR moves `src/librustc/hir/lowering{/, .rs}` to its own crate (`librustc_ast_lowering`) which is very self-contained (only `fn lower_crate` and `trait Resolver` are exposed).

r? @Mark-Simulacrum

### CODE REVIEW

File: src/librustc/hir/ptr.rs [hunk 1]
```diff
@@ -24,8 +24,8 @@ pub fn P<T: 'static>(value: T) -> P<T> {
 
 impl<T: 'static> P<T> {
     // HACK(eddyb) used by HIR lowering in a few places still.
-    // NOTE: do not make this more public than `pub(super)`.
-    pub(super) fn into_inner(self) -> T {
+    // NOTE: do not use outside HIR lowering!
+    pub fn into_inner_only_use_in_lowering(self) -> T {
         *self.ptr
     }
 }
```
Comment (Centril) [inline src/librustc/hir/ptr.rs]: These changes are going away very soon in the 4th arena-ification PR so it shouldn't be an issue.

File: rustfmt.toml [hunk 2]
```diff
@@ -14,7 +14,7 @@ ignore = [
     # to be fixed shortly
     "src/libcore/iter/adapters/mod.rs",
     "src/libcore/iter/traits/iterator.rs",
-    "src/librustc/hir/lowering.rs",
+    "src/librustc_lowering/lib.rs",
```
Comment (Centril) [inline rustfmt.toml]: This got pretty frustrating not being able to do `^C + S` -- would be good to nix the tidy line length check soon so we can fix this.

**Comment by cjgillot** [comment]
I would have tried doing the other way around: moving the HIR definition along with the lowering, as a librustc dependency. 
Why did you prefer this way ?

**Comment by Centril** [comment]
@cjgillot HIR is more entangled into librustc so it will take more work (which I am aiming for in a follow up). But HIR and lowering should, much like the AST and the parser, not be in the same crate.

**Comment by Zoxc** [comment]
@bors r- retry

Won't this heavily conflict with the HIR arena PRs?

**Comment by Mark-Simulacrum** [comment]
I expected no, since it moves the file in a way that git should hopefully understand. But I could be wrong.

**Comment by Centril** [comment]
@Zoxc Yeah I would also expect git to handle this well since it understands the concept of renames. At least that has been my experience in the past when moving files that it has worked well.

**Comment by Zoxc** [comment]
@Centril You can try to rebase https://github.com/rust-lang/rust/pull/66936 on top of this just to confirm it works.

**Comment by petrochenkov** [comment]
FWIW, "lowering" is a pretty generic name applicable to multiple pairs of IRs, while this is about lowering AST into HIR specifically, so I'm not satisfied with the name.

**Comment by Centril** [comment]
@petrochenkov I don't disagree and I'm very open to changing the name, but I also don't have a lot of ideas myself -- perhaps `rustc_ast_lowering` because it transforms AST to HIR.

That said, if we can't find some good name I would appreciate if we could land the PR and bikeshed later as it would be trivial to change the name (since only `rustc_interface` and `rustc_resolve` depend on the crate in uncomplicated ways).

**Comment by petrochenkov** [comment]
>perhaps `rustc_ast_lowering` because it transforms AST to HIR

+1

(Also, for some reason I thought the PR was already merged.)

**Comment by Centril** [comment]
- Last HIR-arena PR has landed.
- I've renamed the crate to `rustc_ast_lowering` per discussion.
- Since this is a new crate, I went with `rustc_span` instead of `syntax_pos` from the get-go.

@bors r= Mark-Simulacrum

**Pull request merged**